SARIF: add partialFingerprints, tags/precision, and ensure absolute Windows paths in artifactLocation.uri

Akindotcome opened the pull request:
This PR improves the SARIF formatter with the following changes:

- Added `partialFingerprints` (`primaryLocationLineHash`) for stable deduplication across runs/refactors.
- Included CWE tags and Bandit test IDs in `tags` for better rule categorization.
- Set `precision` based on Bandit’s confidence levels (HIGH/MEDIUM/LOW → high/medium/low).
- Ensured absolute Windows paths are preserved in `artifactLocation.uri` to match test expectations.
- Added raw `original_path` property for clarity and debugging.

~ All unit tests in `tests/unit/formatters/` pass.  
~ Functional tests are unchanged by this PR.

This should make Bandit’s SARIF output more compliant and interoperable with tools that consume SARIF logs.

Closes #646  
Related to #737

Akindotcome commented:
Hi, I am still hoping my PR gets reviewed

Akindotcome commented:
@sigmavirus24 , @ericwb , @lukehinds 
Hi - friendly ping on this SARIF improvement PR. 

It adds partialFingerprints, tags/precision, preserves Windows absolute paths, and unit tests/unit/formatters/ pass. 

Could one of you kindly review when you have a moment? 

Thanks

ericwb commented on bandit/formatters/sarif.py:
```diff
@@ -12,120 +12,10 @@
 
 This formatter outputs the issues in SARIF formatted JSON.
 
-:Example:
```
Is there a real to remove this rather than just updating with new example output?

The docstring for this formatter and others gets included in our documentation. All of the other formatters have similar examples.

Akindotcome replied:
Thanks for the nudge. I’ve restored the docstring example and updated the output to reflect the new SARIF fields. The formatter’s docstring now matches the rest of Bandit’s formatters and will render correctly in the docs.

Akindotcome replied:
Thanks for catching that, you’re right. I had only restored the Python snippet but missed the JSON output during the git push.

ericwb commented on bandit/formatters/sarif.py:
```diff
@@ -229,45 +121,95 @@ def add_results(issues, run):
+        # Track raw path for run-level properties
+        try:
+            original_paths.add(iss.fname)  # Issue carries 'fname'
+        except Exception:
```
Please use specific exceptions.

ericwb commented on bandit/formatters/sarif.py:
```diff
@@ -229,45 +121,95 @@ def add_results(issues, run):
+        except Exception:
+            try:
+                original_paths.add(iss.as_dict().get("filename", ""))
+            except Exception:
```
Please use specific exceptions.

sigmavirus24 commented on bandit/formatters/sarif.py:
```diff
@@ -12,120 +12,17 @@
-
-"""  # noqa: E501
+"""
+# noqa: E501
```
None of these examples are valid restructuredtext.

gpotter2 commented on bandit/formatters/sarif.py:
```diff
@@ -344,16 +398,36 @@ def create_or_find_rule(issue_dict, rules, rule_indices):
+    # Tags: always include "security"; include CWE tag only if present
+    # and non-zero
+    tags = ["security"]
+    cwe_id = (issue_dict.get("issue_cwe") or {}).get("id")
```
I was wondering if it was just us getting spammed by this user so I went see what other PRs they made and found this. It looks suspiciously AI generated, glad you came to the same conclusion.

https://github.com/secdev/scapy/pull/4844

Akindotcome commented:
@sigmavirus24,  @ericwb 
Thank you for the thorough reviews and feedback. I appreciate the time and detail from everyone.

I’ll address the points you raised (docstring/reST fixes, variable naming consistency, removing the empty except, simplifying confidence/CWE handling, and trimming unnecessary comments) and push the updates immediately I have done the corrections. I’ll be available to iterate further if anything else comes up.

@gpotter2 
Sorry for the misunderstanding here, I am only trying to contribute to meaningful open source projects and grow knowledge-wise as well. Thank you

Akindotcome commented:
Updates:

- Docstring: Converted to valid reStructuredText with `pycon` usage and truncated `json` example. Removed ineffective `noqa`.
- Naming: Renamed `iss` -> `issue` in [add_results()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Fork/bandit/bandit/formatters/sarif.py:185:0-208:30) to minimize diffs.
- Exceptions: Removed empty except; only add `issue.fname` when present.
- Simplifications: [_precision_from_confidence()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Fork/bandit/bandit/formatters/sarif.py:275:0-279:5) uses a direct mapping; CWE extraction via `issue_dict.get("issue_cwe", {}).get("id")`.
- Comments/docstrings: Trimmed redundant comments; tightened helper docstrings.
- Behavior: SARIF structure unchanged (rules/properties/partialFingerprints). [to_uri()](cci:1://file:///c:/Users/brass/OneDrive/Desktop/Fork/bandit/bandit/formatters/sarif.py:415:0-425:41) preserves absolute Windows paths.

Validation:
- `tests/unit/formatters/test_sarif.py` passes locally. Other Windows-local failures are unrelated to this formatter PR and out of scope; CI on Linux/macOS should be unaffected.

sigmavirus24 commented on bandit/formatters/sarif.py:
```diff
@@ -199,18 +192,12 @@ def add_results(issues, run):
-            original_paths.add(getattr(iss, "fname"))
-        else:
-            try:
-                original_paths.add(iss.as_dict().get("filename", ""))
```
Why are we removing this branch?

sigmavirus24 commented on bandit/formatters/sarif.py:
```diff
@@ -401,17 +368,13 @@ def create_or_find_rule(issue_dict, rules, rule_indices):
     help_uri = docs_utils.get_url(rule_id)
 
-    # Precision based on confidence
     precision = _precision_from_confidence(issue_dict.get("issue_confidence"))
```
So this explains your code above. The type-hints above are _wrong_. `confidence` in there is `str | None` in this case. Instead, since this seems to be the usage of that function, it should be `issue_dict.get("issue_confidence", "")` and then you can dramatically simplify the usage of that function. I prefer keeping your current `(str) -> str` signature so let's just change the call site here.

sigmavirus24 reviewed (changes requested):
There is still a lot wrong here.

sigmavirus24 commented on bandit/formatters/sarif.py:
```diff
@@ -199,6 +209,14 @@ def add_results(issues, run):
+            # Fallback to serialized filename if available
+            try:
+                fname2 = issue.as_dict().get("filename")
+            except AttributeError:
```
Why would we encounter an AttributeError here? There should never be a reason for that.

Akindotcome replied:
“Thanks. You’re right.
Since `as_dict()` is guaranteed, there won't be a need for `AttributeError` handling. I'll remove the `try/except` and reuse `fname` for the fallback:

sigmavirus24 commented on bandit/formatters/sarif.py:
```diff
@@ -199,6 +209,14 @@ def add_results(issues, run):
+        else:
+            # Fallback to serialized filename if available
+            try:
+                fname2 = issue.as_dict().get("filename")
```
Why fname2? We can use fname here

ericwb commented on bandit/formatters/sarif.py:
```diff
@@ -10,125 +10,93 @@
-      "runs": [
+   >>> from bandit.formatters import sarif
+   >>> # manager is a BanditManager, tmp is a writable file-like object
+   >>> sarif.report(manager, tmp, 'LOW', 'LOW')
```
The main use of the formatter is using -f sarif from the command line. We don't technically support an API use like provided here. So I'd remove it.

ericwb commented on bandit/formatters/sarif.py:
```diff
@@ -10,125 +10,93 @@
 import logging
 import pathlib
 import sys
+import typing as t
```
Please don't use single character variables or in this case import aliases.

Akindotcome commented on bandit/formatters/sarif.py:
```diff
@@ -10,125 +10,93 @@
-                    ],
-                    "precision": "high"
-                  },
-                  "helpUri": "https://bandit.readthedocs.io/en/1.7.8/plugins/b101_assert_used.html"
```
> What happened to the helpUri?

Placed it in the rule’s reporting descriptor (not per-result)
